File: dd-smoke-tests/play-2.7/app/filters/AbstractFilter.java
package filters;

import akka.stream.Materializer;
import io.opentracing.Scope;
import io.opentracing.Span;
import io.opentracing.Tracer;
import io.opentracing.util.GlobalTracer;
import java.util.concurrent.CompletionStage;
import java.util.function.Function;
import play.libs.concurrent.HttpExecutionContext;
import play.mvc.*;

public abstract class AbstractFilter extends Filter {
  private final HttpExecutionContext ec;
  private final String operationName;
  private final boolean wrap;

  public AbstractFilter(String operationName, Materializer mat, HttpExecutionContext ec) {
    this(operationName, false, mat, ec);
  }

  public AbstractFilter(
      String operationName, boolean wrap, Materializer mat, HttpExecutionContext ec) {
    super(mat);
    this.operationName = operationName;
    this.wrap = wrap;
    this.ec = ec;
  }

  @Override
  public CompletionStage<Result> apply(
      Function<Http.RequestHeader, CompletionStage<Result>> nextFilter,
      Http.RequestHeader requestHeader) {
    final Tracer tracer = GlobalTracer.get();
    final Span startedSpan = wrap ? tracer.buildSpan(operationName).start() : null;
    Scope outerScope = wrap ? tracer.scopeManager().activate(startedSpan) : null;
    try {
      return nextFilter
          .apply(requestHeader)
          .thenApplyAsync(
              result -> {
                Span span = wrap ? startedSpan : tracer.buildSpan(operationName).start();
                try (Scope innerScope = tracer.scopeManager().activate(span)) {
                  // Yes this does no real work
                  return result;
                } finally {
                  span.finish();
                }
              },
              ec.current());
    } finally {
      if (wrap) {
        outerScope.close();
      }
    }
  }
}

File: dd-smoke-tests/play-2.4/app/filters/AbstractFilter.scala
package filters

import io.opentracing.Scope
import io.opentracing.Span
import io.opentracing.Tracer
import io.opentracing.util.GlobalTracer
import play.api.libs.concurrent.Execution.Implicits.defaultContext
import play.api.mvc._
import scala.concurrent.Future

abstract class AbstractFilter(val operationName: String, val wrap: Boolean) extends Filter {
  def this(operationName: String) {
    this(operationName, false)
  }

  override def apply(
      nextFilter: RequestHeader => Future[Result]
  )(requestHeader: RequestHeader): Future[Result] = {
    val tracer      = GlobalTracer.get
    val startedSpan = if (wrap) tracer.buildSpan(operationName).start else null
    val outerScope  =
      if (wrap) tracer.scopeManager.activate(startedSpan) else null
    try {
      nextFilter(requestHeader).map { result =>
        val span =
          if (wrap) startedSpan else tracer.buildSpan(operationName).start
        var innerScope: Scope = null;
        try {
          innerScope = tracer.scopeManager.activate(span)
          // Yes this does no real work
          result
        } finally {
          if (innerScope != null) innerScope.close()
          span.finish
        }
      }
    } finally {
      if (wrap) outerScope.close
    }
  }
}
